src/cargo_communication/request_actor.rs

```diff
@@ -3,39 +3,40 @@ use std::path::{Path, PathBuf};
 use std::process::ExitStatus;
 
 use bsp_server::Message;
-use bsp_server::{RequestId, Response};
+use bsp_server::RequestId;
+pub use cargo_metadata::diagnostic::{
+    Applicability, Diagnostic, DiagnosticCode, DiagnosticLevel, DiagnosticSpan,
+    DiagnosticSpanMacroExpansion,
+};
 use crossbeam_channel::{never, select, Receiver};
-use log::{info, warn};
-use serde_json::to_value;
+use log::warn;
+use mockall::*;
 
 use crate::bsp_types::notifications::{CompileTaskData, MessageType, TaskDataWithKind};
 use crate::bsp_types::requests::Request;
 use crate::bsp_types::StatusCode;
-use crate::cargo_communication::cargo_handle::CargoHandle;
 use crate::cargo_communication::cargo_types::cargo_command::CreateCommand;
 use crate::cargo_communication::cargo_types::cargo_result::CargoResult;
 use crate::cargo_communication::cargo_types::event::{CargoMessage, Event};
 use crate::cargo_communication::request_actor_state::{RequestActorState, TaskState};
 use crate::cargo_communication::utils::get_current_time;
-pub use cargo_metadata::diagnostic::{
-    Applicability, Diagnostic, DiagnosticCode, DiagnosticLevel, DiagnosticSpan,
-    DiagnosticSpanMacroExpansion,
-};
 
-pub struct RequestActor<R>
+pub struct RequestActor<R, C>
 where
     R: Request,
     R::Params: CreateCommand,
     R::Result: CargoResult,
+    C: CargoHandleTrait<CargoMessage>,
 {
     pub(super) sender: Box<dyn Fn(Message) + Send>,
     /// CargoHandle exists to wrap around the communication needed to be able to
     /// run `cargo build/run/test` without blocking. Currently the Rust standard library
     /// doesn't provide a way to read sub-process output without blocking, so we
     /// have to wrap sub-processes output handling in a thread and pass messages
     /// back over a channel.
-    cargo_handle: Option<CargoHandle>,
-    req_id: RequestId,
+    cargo_handle: Option<C>,
+    cancel_receiver: Receiver<Event>,
+    pub(crate) req_id: RequestId,
```

Comment:
Czemu to jest pub(crate) a nie pub(super)?

Reply:
Poprawione